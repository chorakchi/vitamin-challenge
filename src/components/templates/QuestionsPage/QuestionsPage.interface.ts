import { HTMLAttributes } from "react";

export interface QuestionsPageProps extends HTMLAttributes<HTMLElement> {
  data: any;
  title: string;
  description: string;
}
