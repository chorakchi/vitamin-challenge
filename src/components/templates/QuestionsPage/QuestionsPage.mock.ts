

export const mockData = {
  title: "investment simulator",
  description:
    "Questionnaire Questions:",
    data:{
      "EAT": {
          "2017-01-03": 49.169998,
          "2017-02-01": 43.810001,
          "2017-03-01": 42.240002,
          "2017-04-03": 43.790001,
          "2017-05-01": 44.07,
          "2017-06-01": 40.34,
          "2017-07-03": 38.490002,
          "2017-08-01": 35.610001,
          "2017-09-01": 31.24,
          "2017-10-02": 31.5,
          "2017-11-01": 32.900002,
          "2017-12-01": 36.029999,
          "2018-01-02": 39.459999,
          "2018-02-01": 36.939999,
          "2018-03-01": 34.529999,
          "2018-04-02": 35.900002,
          "2018-05-01": 44.869999,
          "2018-06-01": 43.57,
          "2018-07-02": 48.060001,
          "2018-08-01": 44.869999,
          "2018-09-04": 43.639999,
          "2018-10-01": 47.41,
          "2018-11-01": 43.830002,
          "2018-12-03": 50.23,
          "2019-01-02": 44.900002,
          "2019-02-01": 41.360001,
          "2019-03-01": 45.470001,
          "2019-04-01": 44.939999,
          "2019-05-01": 42,
          "2019-06-03": 38.220001,
          "2019-07-01": 38.98,
          "2019-08-01": 40.209999,
          "2019-09-03": 38.139999,
          "2019-10-01": 42.25,
          "2019-11-01": 45.259998,
          "2019-12-02": 45.849998,
          "2020-01-02": 42.279999,
          "2020-02-03": 40.959999,
          "2020-03-02": 34.630001,
          "2020-04-01": 11.49,
          "2020-05-01": 21.879999,
          "2020-06-01": 27.23,
          "2020-07-01": 23.120001,
          "2020-08-03": 25.98,
          "2020-09-01": 45.860001,
          "2020-10-01": 43.009998,
          "2020-11-02": 44.349998,
          "2020-12-01": 51.279999,
          "2021-01-04": 53.580002,
          "2021-02-01": 61.009998,
          "2021-03-01": 72.410004,
          "2021-04-01": 71.910004,
          "2021-05-03": 65.18,
          "2021-06-01": 57.35
      },
      "CAKE": {
          "2017-01-03": 59.43,
          "2017-02-01": 58.98,
          "2017-03-01": 61.189999,
          "2017-04-03": 63.18,
          "2017-05-01": 63.799999,
          "2017-06-01": 59.200001,
          "2017-07-03": 50.009998,
          "2017-08-01": 47.57,
          "2017-09-01": 41.310001,
          "2017-10-02": 42.32,
          "2017-11-01": 45,
          "2017-12-01": 48.029999,
          "2018-01-02": 49.310001,
          "2018-02-01": 50.889999,
          "2018-03-01": 47.099998,
          "2018-04-02": 47.470001,
          "2018-05-01": 52.759998,
          "2018-06-01": 52.040001,
          "2018-07-02": 55.759998,
          "2018-08-01": 49.060001,
          "2018-09-04": 52.889999,
          "2018-10-01": 52.490002,
          "2018-11-01": 48.59,
          "2018-12-03": 47.02,
          "2019-01-02": 43.119999,
          "2019-02-01": 44.970001,
          "2019-03-01": 46.849998,
          "2019-04-01": 48.59,
          "2019-05-01": 49.040001,
          "2019-06-03": 43.889999,
          "2019-07-01": 43.099998,
          "2019-08-01": 44.240002,
          "2019-09-03": 38.700001,
          "2019-10-01": 41.419998,
          "2019-11-01": 41.959999,
          "2019-12-02": 43.939999,
          "2020-01-02": 38.970001,
          "2020-02-03": 37.93,
          "2020-03-02": 36.040001,
          "2020-04-01": 15.72,
          "2020-05-01": 20.129999,
          "2020-06-01": 20.780001,
          "2020-07-01": 21.75,
          "2020-08-03": 23.25,
          "2020-09-01": 30.75,
          "2020-10-01": 28.6,
          "2020-11-02": 29.91,
          "2020-12-01": 38.759998,
          "2021-01-04": 35.41,
          "2021-02-01": 44.959999,
          "2021-03-01": 55.459999,
          "2021-04-01": 59.080002,
          "2021-05-03": 61.89,
          "2021-06-01": 57.23
      },
      "PZZA": {
          "2017-01-03": 85.099998,
          "2017-02-01": 82.790001,
          "2017-03-01": 79.32,
          "2017-04-03": 80.160004,
          "2017-05-01": 78.650002,
          "2017-06-01": 81.959999,
          "2017-07-03": 72,
          "2017-08-01": 71.129997,
          "2017-09-01": 74.419998,
          "2017-10-02": 72.559998,
          "2017-11-01": 62.259998,
          "2017-12-01": 59.959999,
          "2018-01-02": 57.049999,
          "2018-02-01": 63.709999,
          "2018-03-01": 60.040001,
          "2018-04-02": 58.959999,
          "2018-05-01": 63.130001,
          "2018-06-01": 51.259998,
          "2018-07-02": 50.060001,
          "2018-08-01": 42.27,
          "2018-09-04": 45.790001,
          "2018-10-01": 49.27,
          "2018-11-01": 54.759998,
          "2018-12-03": 46.77,
          "2019-01-02": 40.130001,
          "2019-02-01": 38.509998,
          "2019-03-01": 43.32,
          "2019-04-01": 51.900002,
          "2019-05-01": 51.48,
          "2019-06-03": 48.669998,
          "2019-07-01": 44.98,
          "2019-08-01": 43.98,
          "2019-09-03": 46.59,
          "2019-10-01": 52.580002,
          "2019-11-01": 57.98,
          "2019-12-02": 62.82,
          "2020-01-02": 62.799999,
          "2020-02-03": 64.220001,
          "2020-03-02": 57.389999,
          "2020-04-01": 55.73,
          "2020-05-01": 71.57,
          "2020-06-01": 79.639999,
          "2020-07-01": 82.709999,
          "2020-08-03": 95.910004,
          "2020-09-01": 101.610001,
          "2020-10-01": 83.330002,
          "2020-11-02": 75.510002,
          "2020-12-01": 81.279999,
          "2021-01-04": 85.339996,
          "2021-02-01": 101.639999,
          "2021-03-01": 93.790001,
          "2021-04-01": 88.480003,
          "2021-05-03": 97.989998,
          "2021-06-01": 95.389999
      }
  }
};
