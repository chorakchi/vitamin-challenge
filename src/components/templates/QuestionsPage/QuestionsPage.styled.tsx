import styled from "styled-components";

export const QuestionsPageStyled = styled.div`
padding: 5%;
margin: 0;
height: 100%;
width: 100%;
display: grid;
grid-column: 1;
grid-auto-flow: column;
grid-template-columns: max-content;
gap: 128px;

.mainBlock{
    width: 100%;
    height: 100%;
    gap: 18px;
}
`;
