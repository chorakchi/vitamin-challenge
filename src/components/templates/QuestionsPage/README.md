this template component display information in a chart view from the information entry in form. 

### Questionnaire Questions:

- What’s your name?
- How much do you earn per month?
- What is your risk tolerance level? (Slider between 0-10)

Once the questionnaire is completed, she should see a dashboard showing a graph of how her portfolio has developed between 1.1.2017 and 1.6.2021.

The portfolio develops as follows:

- At the beginning of each month (the first day that the stock market is open), Lea always invests 15% of her monthly income. The stocks she invests in are those in her portfolio, based on the supplied weighting.
- Assume that Lea’s first contribution took place on 3.1.2017 and her last contribution on 1.6.2021
- The close price should be used as the purchase price of the partial share

---

### **Worked Example**

As an example, with a from date of 1.1.2017 and to date of 3.6.2021, if Lea earns $3000 per month and has a risk tolerance of 5

- Her monthly contribution would be $450

Therefore on 3.1.2017 she invested:

- $292.50 in *CAKE for 4.92 shares*
- $90.00 in *PZZA for 1.06 shares*
- $67.50 in *EAT for 1.37 shares*

If she continued to invest monthly, as of 1.6.2021, these shares would be worth:

- CAKE: ~$22,059.73 (~385.46 shares)
- PZZA: ~$7,457.61 (~78.18 shares)
- EAT: ~$5,393.25 (~94.04 shares)

> you can try it out manually, it working with moke data copied from API.

### Playground:
