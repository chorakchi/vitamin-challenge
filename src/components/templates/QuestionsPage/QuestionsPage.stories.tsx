import { ComponentStory, ComponentMeta } from "@storybook/react";
import { mockData } from "./QuestionsPage.mock";
import { QuestionsPage } from "./QuestionsPage";
// @ts-ignore
import readme from "./README.md";

export default {
  title: "templates/QuestionsPage",
  component: QuestionsPage,
  parameters: {
    backgrounds: {
      default: "light",
    },
    docs: {
      description: {
        component: readme,
      },
    },
  },
} as ComponentMeta<typeof QuestionsPage>;

const Template: ComponentStory<typeof QuestionsPage> = (args) => <QuestionsPage {...args} />;

export const QuestionsPage_ = Template.bind({});
QuestionsPage_.args = mockData;
