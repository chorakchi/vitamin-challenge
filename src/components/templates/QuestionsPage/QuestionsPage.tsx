import { QuestionsPageProps } from "./QuestionsPage.interface";
import { PageLayout, Questions, Chart, Stack, Typography } from "components";
import { useCallback, useState } from "react";
import useInvestment from "helpers/investment";
import { shares } from "helpers/shares";

// Create our number formatter.
var formatter = (num) => (num.toLocaleString('en-US', {
  style: 'currency',
  currency: 'USD',
}))

export const QuestionsPage: React.FC<QuestionsPageProps> = ({
  data,
  description,
  ...props
}) => {
  const [name, setName] = useState("");
  const [earn, setEarn] = useState();
  const [risk, setRisk] = useState(6);
  const shareVales = shares({
    risk: risk,
    contribution: ((earn || 0) / 100) * 15,
  });
  const [result,calcInvestment] = useInvestment();

  const onChangeName = useCallback(
    (e) => {
        // Set the component's new state based on the user's input
        setName(e.target.value);
    },
    [
        /* necessary deps */
        name,
    ]
);

const onChangeEarn = useCallback(
  ({ target }) => {
      // Set the component's new state based on the user's input
      setEarn(target.value);
      calcInvestment({ data, shareVales })
  },
  [
      /* necessary deps */
      earn,
  ]
);

const onChangeRisk = useCallback(
  ({ target }) => {
      // Set the component's new state based on the user's input
      setRisk(target.value);
      calcInvestment({ data, shareVales })
  },
  [
      /* necessary deps */
      risk,
  ]
);

  return (
    <PageLayout {...props} description={<>{description}</>}>
      <Stack direction="column">
        <Questions
          nameProps={{
            onChange: onChangeName
          }}
          earnProps={{
            onChange: onChangeEarn
          }}
          riskProps={{
            onChange: onChangeRisk
          }}
        />
        {earn && (
          <>
            <Typography bold>Risk Level: {risk}</Typography>
            <Typography bold>
              Monthly contribution: {(((earn || 0) / 100) * 15).toFixed(2)}
            </Typography>
            <br />
            {result.map((item) => (
              <Typography>
                {item.name + " : " + formatter(item.data[item.data.length - 1][1])}
              </Typography>
            ))}
          </>
        )}
      </Stack>
      {earn ? <Chart items={result} /> : ""}
    </PageLayout>
  );
};
