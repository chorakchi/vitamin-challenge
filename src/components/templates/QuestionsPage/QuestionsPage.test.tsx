import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { QuestionsPage } from "./QuestionsPage";
import { mockData } from "./QuestionsPage.mock";

describe("QuestionsPage", () => {
  it("should render the QuestionsPage", () => {
    const subject = shallow(
      <QuestionsPage {...mockData}
      />
    );
    expect(subject).toHaveLength(1);
  });

  it("should render correctly and match to the Snapshot", () => {
    const subject = renderer
      .create(
        <QuestionsPage {...mockData}
        />
      )
      .toJSON();
    expect(subject).toMatchSnapshot();
  });
});
