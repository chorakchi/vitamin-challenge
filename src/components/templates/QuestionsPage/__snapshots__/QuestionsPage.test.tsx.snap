// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionsPage should render correctly and match to the Snapshot 1`] = `
<div
  className="PageLayoutstyled__StyledWrapper-sc-1kudg7l-0 bZBWEP"
>
  <div
    className="Stackstyled__StackStyled-sc-1adxq1u-0 bnrRyH"
    direction="column"
  >
    <div
      className="Typographystyled__TypographyStyled-sc-i22hyo-0 jzzjid"
    >
      investment simulator
    </div>
    <div
      className="Typographystyled__TypographyStyled-sc-i22hyo-0 iKdygM"
    >
      Questionnaire Questions:
    </div>
    <div
      className="Stackstyled__StackStyled-sc-1adxq1u-0 jWmaRX mainBlock"
      direction="row"
    >
      <div
        className="Stackstyled__StackStyled-sc-1adxq1u-0 bnrRyH"
        direction="column"
      >
        <div
          className="Questionsstyled__StyledWrapper-sc-1yeop21-0 cNkGxQ"
        >
          <div
            className="Stackstyled__StackStyled-sc-1adxq1u-0 bnrRyH"
            direction="column"
          >
            <label
              className="Typographystyled__TypographyStyled-sc-i22hyo-0 iKdygM"
            >
              Whatâ€™s your name?
              <br />
              <input
                className="TextFieldstyled__TextFieldStyled-sc-mvtmc2-0 NTHeK"
                onChange={[Function]}
                type="text"
              />
            </label>
          </div>
          <div
            className="Stackstyled__StackStyled-sc-1adxq1u-0 bnrRyH"
            direction="column"
          >
            <label
              className="Typographystyled__TypographyStyled-sc-i22hyo-0 iKdygM"
            >
              How much do you earn per month?
              <br />
              <input
                className="TextFieldstyled__TextFieldStyled-sc-mvtmc2-0 NTHeK"
                onChange={[Function]}
                type="number"
              />
            </label>
          </div>
          <span>
            <label
              className="Typographystyled__TypographyStyled-sc-i22hyo-0 iKdygM"
            >
              What is your risk tolerance level?
              <br />
              <input
                className="Sliderstyled__SliderStyled-sc-fj8kj3-0 ktJaBw"
                max="10"
                min="2"
                onChange={[Function]}
                step="2"
                type="range"
              />
            </label>
          </span>
        </div>
      </div>
      
    </div>
  </div>
</div>
`;
